per-resource options fail with pmi=pmix inside Slurm allocation

This could very well be related to #129 but I'm going to open a new issue as the situation is slightly different and I don't want to confuse things if they end up being separate root causes. In particular, as the title implies, I am only seeing issues when I use the [per-resource options to flux run](https://flux-framework.readthedocs.io/projects/flux-core/en/latest/man1/flux-run.html#per-resource-options), particularly `--tasks-per-node=N` or `--tasks-per-core=N` (I'm not currently testing anything with GPUs, and just `--cores=N` works fine).

I am trying with both OpenMPI and ParaStationMPI (which is an MPICH derivative, but also uses PMIx).

## ParaStationMPI

Starting with ParaStationMPI, the startup looks like:
<details><summary>`srun flux start` log with ParaStationMPI</summary>

```shell
$ srun -c ${SLURM_JOB_CPUS_PER_NODE%%(*} --pty flux start
boot_pmi: flux-pmi-client: trying 'simple'
boot_pmi: simple: PMI_FD not found in environ
boot_pmi: flux-pmi-client: trying 'pmix'
boot_pmi: pmix: selected
boot_pmi: pmix: initialize: rank=0 size=2 name=pspmix_0x00010001e1bc[1:123324]: success
boot_pmi: pmix: get key=flux.instance-level: PMIx_Get: PMIX_ERR_NOT_FOUND
boot_pmi: pmix: get key=flux.taskmap: PMIx_Get: PMIX_ERR_NOT_FOUND
boot_pmi: pmix: get key=PMI_process_mapping: PMIx_Get: PMIX_ERR_NOT_FOUND
boot_pmi: pmix: put key=0 value={"host":"jrc0001","pubkey":"C.MM0>y#50WUc@ZBq4DQCxpYd03oBQ@LBqug]!Me","uri":["tcp://10.14.1.11:34417"]}: success
boot_pmi: pmix: barrier: success
boot_pmi: pmix: get key=1 value={"host":"jrc0003","pubkey":"fVIhVSdMOO+g?8Ri&s+>mml!m@a-KO^#}c80sPR>","uri":[]}: success
boot_pmi: pmix: get key=0 value={"host":"jrc0001","pubkey":"C.MM0>y#50WUc@ZBq4DQCxpYd03oBQ@LBqug]!Me","uri":["tcp://10.14.1.11:34417"]}: success
boot_pmi: pmix: barrier: success
boot_pmi: pmix: finalize: success
```
</details>

And then if I run a simple MPI application either without specifying `-o pmi=pmix` using only "per-task" options everything works as expected:
<details><summary>working `flux run` logs with ParaStationMPI</summary>

```shell
$ flux run -N1 -n2 --label-io ./mpi-hello-world
1: Hello world from processor jrc0003, rank 1 out of 2 processors
0: Hello world from processor jrc0003, rank 0 out of 2 processors

$ flux run -N1 --tasks-per-node=2 --label-io ./mpi-hello-world
1: Hello world from processor jrc0001, rank 1 out of 2 processors
0: Hello world from processor jrc0001, rank 0 out of 2 processors

$ flux run -N1 -n2 --label-io -o pmi=pmix ./mpi-hello-world
1: Hello world from processor jrc0003, rank 1 out of 2 processors
0: Hello world from processor jrc0003, rank 0 out of 2 processors
```
</details>

But as soon as I combine `-o pmi=pmix` with `--tasks-per-node=N` for N>1 then I get problems (it still works for N=1, even with multiple nodes...)
<details><summary>failing `flux run` log with ParaStationMPI (with extra shell verbosity)</summary>

```shell
$ flux run -N1 --tasks-per-node=2 --label-io -o verbose=2 -o pmi=pmix ./mpi-hello-world
0.071s: flux-shell[0]: DEBUG: Loading /p/software/default/stages/2026/software/Flux/0.80.0-GCCcore-14.3.0/etc/flux/shell/initrc.lua
0.072s: flux-shell[0]: TRACE: Successfully loaded flux.shell module
0.072s: flux-shell[0]: TRACE: trying to load /p/software/default/stages/2026/software/Flux/0.80.0-GCCcore-14.3.0/etc/flux/shell/initrc.lua
0.076s: flux-shell[0]: DEBUG: pmix: server is enabled
0.077s: flux-shell[0]: TRACE: trying to load /p/software/default/stages/2026/software/Flux/0.80.0-GCCcore-14.3.0/etc/flux/shell/lua.d/openmpi.lua
0.092s: flux-shell[0]: DEBUG: output.kvs: batch timeout = 0.500s
0.098s: flux-shell[0]: DEBUG: oom: monitoring /sys/fs/cgroup/slurm/user-maloney2/job-15416630/step-0/tasks/memory.events
0.098s: flux-shell[0]: DEBUG: rexec: registered rexec service as 23957-shell-foqTX4Ws.rexec
0.098s: flux-shell[0]: DEBUG: pmix: jobid = 1782965075968
0.098s: flux-shell[0]: DEBUG: pmix: shell_rank = 0
0.098s: flux-shell[0]: DEBUG: pmix: local_nprocs = 2
0.098s: flux-shell[0]: DEBUG: pmix: total_nprocs = 1
0.098s: flux-shell[0]: DEBUG: pmix: server outsourced to OpenPMIx 5.0.8 (PMIx Standard: 5.1
0.222s: flux-shell[0]: DEBUG: pmix: local_peers = 0,1
0.222s: flux-shell[0]: DEBUG: pmix: node_map = jrc0003
0.222s: flux-shell[0]: DEBUG: pmix: proc_map = 0,1
0.223s: flux-shell[0]: DEBUG: 0: task_count=2 slot_count=1 cores_per_slot=128 slots_per_node=1
0.223s: flux-shell[0]: DEBUG: 0: tasks [0-1] on cores 0-127
0.224s: flux-shell[0]: DEBUG: hwloc: Writing 154259 bytes to HWLOC_XMLFILE=/tmp/flux-hQypsl/jobtmp-0-ƒoqTX4Ws/hwloc.xml
0.506s: flux-shell[0]: DEBUG: task 1 complete status=1
1: <PSP:r0000001:Backtrace after SIGSEGV (Invalid memory reference):>
1: <PSP:r0000001:# 0: /p/software/default/stages/2026/software/pscom/6-default-GCCcore-14.3.0/lib/libpscom.so.5(+0xdb91) [0x14f6e0206b91]>
1: <PSP:r0000001:# 1: /usr/lib64/libc.so.6(+0x3fc60) [0x14f6e11e1c60]>
1: <PSP:r0000001:# 2: /usr/lib64/libc.so.6(+0x15b409) [0x14f6e12fd409]>
1: <PSP:r0000001:# 3: /p/software/default/stages/2026/software/psmpi/5.13.0-1-GCC-14.3.0/lib/libmpi.so.12(+0x2fb7b3) [0x14f6e16ac7b3]>
1: <PSP:r0000001:# 4: /p/software/default/stages/2026/software/psmpi/5.13.0-1-GCC-14.3.0/lib/libmpi.so.12(+0x2ef1cb) [0x14f6e16a01cb]>
1: <PSP:r0000001:# 5: /p/software/default/stages/2026/software/psmpi/5.13.0-1-GCC-14.3.0/lib/libmpi.so.12(+0x29c607) [0x14f6e164d607]>
1: <PSP:r0000001:# 6: /p/software/default/stages/2026/software/psmpi/5.13.0-1-GCC-14.3.0/lib/libmpi.so.12(+0x29ca05) [0x14f6e164da05]>
1: <PSP:r0000001:# 7: /p/software/default/stages/2026/software/psmpi/5.13.0-1-GCC-14.3.0/lib/libmpi.so.12(MPI_Init+0x30) [0x14f6e1531440]>
1: <PSP:r0000001:# 8: ./mpi-hello-world() [0x4011a7]>
1: <PSP:r0000001:# 9: /usr/lib64/libc.so.6(+0x2a610) [0x14f6e11cc610]>
1: <PSP:r0000001:#10: /usr/lib64/libc.so.6(__libc_start_main+0x80) [0x14f6e11cc6c0]>
1: <PSP:r0000001:#11: ./mpi-hello-world() [0x4010b5]>
0: Hello world from processor jrc0003, rank 0 out of 1 processors
0.550s: flux-shell[0]: DEBUG: task 0 complete status=0
0.561s: flux-shell[0]: DEBUG: exit 1
flux-job: task(s) exited with exit code 1
```
</details>

and adding the `--env=UCX_LOG_LEVEL=trace` option gives  [ucx_trace.log](https://github.com/user-attachments/files/29884937/ucx_trace.log)

Checking the UCX bandwidth as suggested in #129 gives:
<details><summary>UCX bandwidth with ParaStationMPI</summary>

```shell
$ flux run --label-io -N2 -n2 sh -c 'ucx_info -d | grep -B3 -i bandwidth' 
1: #  System device: <unknown>
1: #
1: #      capabilities:
1: #            bandwidth: 0.00/ppn + 19360.00 MB/sec
1: --
1: #  System device: enp225s0f0 (0)
1: #
1: #      capabilities:
1: #            bandwidth: 113.16/ppn + 0.00 MB/sec
1: --
1: #  System device: ib0 (1)
1: #
1: #      capabilities:
1: #            bandwidth: 2200.00/ppn + 0.00 MB/sec
1: --
1: #  System device: <unknown>
1: #
1: #      capabilities:
1: #            bandwidth: 11.91/ppn + 0.00 MB/sec
1: --
1: #  System device: <unknown>
1: #
1: #      capabilities:
1: #            bandwidth: 0.00/ppn + 15360.00 MB/sec
1: --
1: #  System device: <unknown>
1: #
1: #      capabilities:
1: #            bandwidth: 0.00/ppn + 15360.00 MB/sec
1: --
1: #  System device: mlx5_0 (1)
1: #
1: #      capabilities:
1: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
1: --
1: #  System device: mlx5_0 (1)
1: #
1: #      capabilities:
1: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
1: --
1: #  System device: mlx5_0 (1)
1: #
1: #      capabilities:
1: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
1: --
1: #  System device: mlx5_0 (1)
1: #
1: #      capabilities:
1: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
1: --
1: #  System device: mlx5_0 (1)
1: #
1: #      capabilities:
1: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
1: --
1: #  System device: <unknown>
1: #
1: #      capabilities:
1: #            bandwidth: 0.00/ppn + 11145.00 MB/sec
0: #  System device: <unknown>
0: #
0: #      capabilities:
0: #            bandwidth: 0.00/ppn + 19360.00 MB/sec
0: --
0: #  System device: enp225s0f0 (0)
0: #
0: #      capabilities:
0: #            bandwidth: 113.16/ppn + 0.00 MB/sec
0: --
0: #  System device: ib0 (1)
0: #
0: #      capabilities:
0: #            bandwidth: 2200.00/ppn + 0.00 MB/sec
0: --
0: #  System device: <unknown>
0: #
0: #      capabilities:
0: #            bandwidth: 11.91/ppn + 0.00 MB/sec
0: --
0: #  System device: <unknown>
0: #
0: #      capabilities:
0: #            bandwidth: 0.00/ppn + 15360.00 MB/sec
0: --
0: #  System device: <unknown>
0: #
0: #      capabilities:
0: #            bandwidth: 0.00/ppn + 15360.00 MB/sec
0: --
0: #  System device: mlx5_0 (1)
0: #
0: #      capabilities:
0: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
0: --
0: #  System device: mlx5_0 (1)
0: #
0: #      capabilities:
0: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
0: --
0: #  System device: mlx5_0 (1)
0: #
0: #      capabilities:
0: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
0: --
0: #  System device: mlx5_0 (1)
0: #
0: #      capabilities:
0: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
0: --
0: #  System device: mlx5_0 (1)
0: #
0: #      capabilities:
0: #            bandwidth: 11794.23/ppn + 0.00 MB/sec
0: --
0: #  System device: <unknown>
0: #
0: #      capabilities:
0: #            bandwidth: 0.00/ppn + 11145.00 MB/sec
```
</details>

## OpenMPI

Switching to OpenMPI, the startup looks like:
<details><summary>`srun flux start` log with OpenMPI</summary>

```shell
$ srun -c ${SLURM_JOB_CPUS_PER_NODE%%(*} --pty flux start
boot_pmi: flux-pmi-client: trying 'simple'
boot_pmi: simple: PMI_FD not found in environ
boot_pmi: flux-pmi-client: trying 'pmix'
boot_pmi: pmix: selected
boot_pmi: pmix: initialize: rank=0 size=2 name=pspmix_0x02730000aa3d[627:43581]: success
boot_pmi: pmix: get key=flux.instance-level: PMIx_Get: PMIX_ERR_NOT_FOUND
boot_pmi: pmix: get key=flux.taskmap: PMIx_Get: PMIX_ERR_NOT_FOUND
boot_pmi: pmix: get key=PMI_process_mapping: PMIx_Get: PMIX_ERR_NOT_FOUND
boot_pmi: pmix: put key=0 value={"host":"jrc0627","pubkey":"^UW4Xg3]KZ98L8JO{cuf9jodk3T{S+f6BS*7CxPo","uri":["tcp://10.14.4.45:34443"]}: success
boot_pmi: pmix: barrier: success
boot_pmi: pmix: get key=1 value={"host":"jrc0628","pubkey":"QLZbI&Kc(yNWHUs+36w6+ge%xj[*XUGHU7a2uZZY","uri":[]}: success
boot_pmi: pmix: get key=0 value={"host":"jrc0627","pubkey":"^UW4Xg3]KZ98L8JO{cuf9jodk3T{S+f6BS*7CxPo","uri":["tcp://10.14.4.45:34443"]}: success
boot_pmi: pmix: barrier: success
boot_pmi: pmix: finalize: success
```
</details>

With OpenMPI I have to specify `-o pmi=pmix` to `flux run` otherwise I only get singletons. If I then use only "per-task" options everything works as expected:
<details><summary>working `flux run` logs with OpenMPI</summary>

```shell
$ flux run -N1 -n2 --label-io -o pmi=pmix ./mpi-hello-world
1: Hello world from processor jrc0628, rank 1 out of 2 processors
0: Hello world from processor jrc0628, rank 0 out of 2 processors
```
</details>

But with `--tasks-per-node=N` for N>1 I get the following errors:
<details><summary>failing `flux run` log with OpenMPI</summary>

```shell
$ flux run -N1 --tasks-per-node=2 --label-io -o pmi=pmix ./mpi-hello-world
1: *** An error occurred in MPI_Init
1: *** on a NULL communicator
1: *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
1: ***    and MPI will try to terminate your MPI job as well)
1: [jrc0628:43678] [0] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libopen-pal.so.80(opal_backtrace_buffer+0x20) [0x147b4f9b8830]
1: [jrc0628:43678] [1] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libmpi.so.40(ompi_mpi_abort+0x11f) [0x147b4ff932ff]
1: [jrc0628:43678] [2] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libmpi.so.40(ompi_mpi_errors_are_fatal_comm_handler+0xb9) [0x147b4ff83609]
1: [jrc0628:43678] [3] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libmpi.so.40(ompi_errhandler_invoke+0xc3) [0x147b4ff82863]
1: [jrc0628:43678] [4] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libmpi.so.40(MPI_Init+0xab) [0x147b4ffc28fb]
1: [jrc0628:43678] [5] func:./mpi-hello-world() [0x4011a7]
1: [jrc0628:43678] [6] func:/usr/lib64/libc.so.6(+0x2a610) [0x147b4fd18610]
1: [jrc0628:43678] [7] func:/usr/lib64/libc.so.6(__libc_start_main+0x80) [0x147b4fd186c0]
1: [jrc0628:43678] [8] func:./mpi-hello-world() [0x4010b5]
1: [jrc0628:43678] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
1.317s: flux-shell[0]:  WARN: pmix: notify source=ƒg5jcvPy.1 event-status=-61
2.483s: job.exception ƒg5jcvPy type=exec severity=0 ƒg5jcvPy.0 called PMIx_Abort (status=14): (no message)
0: [jrc0628:43677] pml_ucx.c:188  Error: Failed to receive UCX worker address: Take next option (-46)
0: [jrc0628:43677] [0] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libopen-pal.so.80(opal_backtrace_buffer+0x20) [0x153de4b1b830]
0: [jrc0628:43677] [1] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libmpi.so.40(ompi_mpi_abort+0x11f) [0x153de50f62ff]
0: [jrc0628:43677] [2] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libmpi.so.40(ompi_mpi_errors_are_fatal_comm_handler+0xb9) [0x153de50e6609]
0: [jrc0628:43677] [3] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libmpi.so.40(ompi_errhandler_invoke+0xc3) [0x153de50e5863]
0: [jrc0628:43677] [4] func:/p/software/default/stages/2026/software/OpenMPI/5.0.8-GCC-14.3.0/lib/libmpi.so.40(MPI_Init+0xab) [0x153de51258fb]
0: [jrc0628:43677] [5] func:./mpi-hello-world() [0x4011a7]
0: [jrc0628:43677] [6] func:/usr/lib64/libc.so.6(+0x2a610) [0x153de4e7b610]
0: [jrc0628:43677] [7] func:/usr/lib64/libc.so.6(__libc_start_main+0x80) [0x153de4e7b6c0]
0: [jrc0628:43677] [8] func:./mpi-hello-world() [0x4010b5]
flux-job: task(s) Terminated
```
</details>

and adding the `--env=UCX_LOG_LEVEL=trace` option gives (I also tried `--env=OMPI_MCA_pml_base_verbose=100` but it didn't seem to produce any further output)  [ucx_trace_ompi.log](https://github.com/user-attachments/files/29885673/ucx_trace_ompi.log)

The UCX bandwidth info is unchanged.

--- a/src/shell/plugins/main.c
+++ b/src/shell/plugins/main.c
@@ -207,9 +207,10 @@ static int px_init (flux_plugin_t *p,
     px->shell = shell;
 
     if (flux_shell_info_unpack (shell,
-                                "{s:I s:i}",
+                                "{s:I s:i s:i}",
                                 "jobid", &px->id,
-                                "rank", &px->shell_rank) < 0)
+                                "rank", &px->shell_rank,
+                                "ntasks", &px->total_nprocs) < 0)
         return -1;
     shell_debug ("jobid = %ju", (uintmax_t)px->id);
     shell_debug ("shell_rank = %d", px->shell_rank);
@@ -222,11 +223,6 @@ static int px_init (flux_plugin_t *p,
                                      &px->local_nprocs) < 0)
         return -1;
     shell_debug ("local_nprocs = %d", px->local_nprocs);
-    if (flux_shell_jobspec_info_unpack (shell,
-                                        "{s:i}",
-                                        "ntasks",
-                                        &px->total_nprocs) < 0)
-        return -1;
     shell_debug ("total_nprocs = %d", px->total_nprocs);
     if (!(px->job_tmpdir = flux_shell_getenv (shell, "FLUX_JOB_TMPDIR")))
         return -1;

--- a/t/t0002-basic.t
+++ b/t/t0002-basic.t
@@ -73,6 +73,33 @@ test_expect_success '2n4p pmix.univ.size is set correctly' '
 			| sort -n >pmix.univ.size.out &&
 	test_cmp pmix.univ.size.exp pmix.univ.size.out
 '
+# issue #130: per-resource placement (--tasks-per-node) expands the task
+# count via a shell option, not the jobspec.  The job size must reflect the
+# expanded total; a stale value here previously broke MPI wireup.
+test_expect_success '2n4p pmix.job.size is correct with --tasks-per-node' '
+	cat >tpn.pmix.job.size.exp <<-EOT &&
+	0: 4
+	1: 4
+	2: 4
+	3: 4
+	EOT
+	run_timeout 30 flux run -N2 --tasks-per-node=2 \
+		${GETKEY} --proc=* --label-io pmix.job.size \
+			| sort -n >tpn.pmix.job.size.out &&
+	test_cmp tpn.pmix.job.size.exp tpn.pmix.job.size.out
+'
+test_expect_success '2n4p pmix.univ.size is correct with --tasks-per-node' '
+	cat >tpn.pmix.univ.size.exp <<-EOT &&
+	0: 4
+	1: 4
+	2: 4
+	3: 4
+	EOT
+	run_timeout 30 flux run -N2 --tasks-per-node=2 \
+		${GETKEY} --proc=* --label-io pmix.univ.size \
+			| sort -n >tpn.pmix.univ.size.out &&
+	test_cmp tpn.pmix.univ.size.exp tpn.pmix.univ.size.out
+'
 test_expect_success '2n3p pmix.local.size is set correctly' '
 	cat >2n3p.pmix.local.size.exp <<-EOT &&
 	0: 2

--- a/t/t1000-ompi-basic.t
+++ b/t/t1000-ompi-basic.t
@@ -49,6 +49,28 @@ test_expect_success '2n2p ompi hello' '
 # --env=OMPI_MCA_btl_base_verbose=100 \
 # --env=OMPI_MCA_btl_tcp_if_exclude=docker0,lo \
 
+# see issue #130
+# per-resource task placement (--tasks-per-node, --tasks-per-core) expands
+# the task count via a shell option rather than the jobspec, so the pmix
+# plugin must source the job size from shell info, not jobspec info.
+test_expect_success '1n2p ompi hello with --tasks-per-node' '
+	run_timeout 30 flux run -N1 --tasks-per-node=2 \
+		${MPI_HELLO} >hello_tpn_1n2p.out &&
+	grep "There are 2 tasks" hello_tpn_1n2p.out
+'
+
+test_expect_success '2n4p ompi hello with --tasks-per-node' '
+	run_timeout 30 flux run -N2 --tasks-per-node=2 \
+		${MPI_HELLO} >hello_tpn_2n4p.out &&
+	grep "There are 4 tasks" hello_tpn_2n4p.out
+'
+
+test_expect_success '1n2p ompi hello with --tasks-per-core' '
+	run_timeout 30 flux run -N1 --tasks-per-core=2 --cores=1 \
+		${MPI_HELLO} >hello_tpc_1n2p.out &&
+	grep "There are 2 tasks" hello_tpc_1n2p.out
+'
+
 # see issue #27
 test_expect_success '2n3p ompi hello doesnt hang' '
 	run_timeout 60 flux run -N2 -n3 \

